update platform linux68k

tevorbl opened the pull request:
latest version of musashi engine
includes floating point emulation
(plus a few patches to add in missing opcodes needed by ack - see tags JFF & TBB)

added a few missing linux syscalls in sim.c

pascal now runs pretty well
quick test with modula2 passes

c gets the floating point numbers wrong, so more work needed here.  Basic hello world type programs seem OK, though.

other languages untested

plat/linux68k/emu/build.lua is probably not quite right - the softfloat directory is compiled in the wrong place.   Any help here would be appreciated.

tevorbl committed:
update platform linux68k

latest version of musashi engine
includes floating point emulation
(plus a few patches to add in missing opcodes needed by ack - see tags JFF & TBB)

added a few missing linux syscalls in sim.c

pascal now runs pretty well
quick test with modula2 passes

c gets the floating point numbers wrong, so more work needed here

other languages untested

plat/linux68k/emu/build.lua is probably not quite right - the softfloat directory is compiled in the wrong place

tevorbl committed:
fpu bug fix for fgetmant op

- inserted missing break
- use temp variable instead of manipulating dest register

fgetexp changed to directly convert int to floatx80

tevorbl committed:
fixed floats in printf in linux59k platform

problem was implementation of fint & fintrz floating point ops in m68kfpu.c
Both these ops truncated the integer to 32 bits instead of leaving it as an extended precision floating point number

Changed the implementation to use 64 bit ints instead of 32 bit ints.

tevorbl committed:
another fix for printf(float) on m68k platform

bug caused by this instruction:

  fmove.l fp0,d0 

problem was caused by a conflict between the fpu emulator (softfloat) and the compiler.

the emulator implemented this as a purely arithmetic move & conversion,
but the compiler assumed that the result could be interpreted as a logical (ie unsigned) conversion.
rightly or wrongly.


for example, if fp0 contained the value 2576980377.0  which is unsigned integer -1717987328
the emulator would treat this as an integer overflow and move 0x7fffffff (INT_MAX) into d0.

The complier on the other hand would assume that d0 contained 2576980377 (the unsigned value).

I don't know which is correct, but this is my fix for the time being.

davidgiven commented on plat/linux68k/emu/sim.c:
```diff
@@ -310,11 +361,15 @@ int main(int argc, char* argv[])
-	m68k_set_cpu_type(M68K_CPU_TYPE_68020);
-	m68k_init();
-	m68k_pulse_reset();
+        //printf("now at line %d\n", __LINE__);
```
stray tracing

tevorbl replied:
OK, should be done now.
formatting is leading tabs, stray tracing removed.
Also restored temp dir & CC in top level Makefile

davidgiven commented on Makefile:
```diff
@@ -9,7 +9,8 @@ DEFAULT_PLATFORM = pc86
 
-ACK_TEMP_DIR = /tmp
+#ACK_TEMP_DIR = /tmp
+ACK_TEMP_DIR = tmp
```
please revert this file

davidgiven commented:
Thank you very much for this! Particularly as I was halfway through explaining how I couldn't use it because new the license had a non-commercial use clause before realising I had the patch backwards and the _old_ version had the non-commercial use clause. Er, yikes. So, er, definitely thanks.

I reckon it's pretty much good to go, apart from a couple of really nitpicky comments about stray comments and debugging changes.

(I actually have another compiler project which could really use this, too...)

tevorbl committed:
minor formatting changes

sim.c:
remove stray tracing statements
make formatting consistently leading tabs

Makefile:
restore ACK_TEMP_DIR & CC to default settings

davidgiven commented:
Fantastic, thanks! Merging.
